web/src/composebox_typeahead.ts

```diff
@@ -1179,6 +1191,56 @@ export function content_typeahead_selected(
     return beginning + rest;
 }
 
+function will_produce_broken_stream_topic_link(word: string): boolean {
+    return invalid_stream_topic_regex.test(word);
+}
+
+function get_stream_name(syntax: string): string {
+    const start = syntax.lastIndexOf("#**");
+    const end = syntax.lastIndexOf(">");
+    return syntax.slice(start + 3, end);
+}
+export function replace_markdown_characters(text: string): string {
+    // Replace markdown characters with their escaped versions.
+    return text.replaceAll(invalid_stream_topic_regex, (match) => {
+        switch (match) {
+            case "`":
+                return "&grave;";
+            case ">":
+                return "&gt;";
+            case "*":
+                return "&#42;";
+            case "$$":
+                return "&#36;&#36;";
+            default:
+                return match;
+        }
+    });
+}
+
+function url_syntax(stream_name: string, topic_name: string | undefined): string {
+    const stream = stream_data.get_sub(stream_name);
+    const stream_id = stream?.stream_id;
+    assert(stream_id !== undefined);
+    if (topic_name !== undefined) {
+        return `[#${replace_markdown_characters(stream_name)}>${replace_markdown_characters(topic_name)}](${internal_url.by_stream_topic_url(stream_id, topic_name, () => stream_name)}) `;
+    }
+    return `[#${replace_markdown_characters(stream_name)}](${internal_url.by_stream_url(stream_id, () => stream_name)}) `;
+}
+
+export function stream_topic_link_syntax(typed_syntax_text: string, topic_name: string): string {
+    const stream_name = get_stream_name(typed_syntax_text);
+    // If the topic name is such that it will generate an invalid #**stream>topic** syntax,
+    // we revert to generating the normal markdown syntax for a link.
+    if (
+        will_produce_broken_stream_topic_link(topic_name) ||
+        will_produce_broken_stream_topic_link(stream_name)
+    ) {
+        return url_syntax(stream_name, topic_name);
+    }
+    return `#**${stream_name}>${topic_name}** `;
```

Comment:
Why is there a trailing space here? Seems like it should be added by the caller if desired.

Reply:
I noticed in the module that a trailing space was always put after inserting any sort of syntax, so I included it here itself to keep the caller clean.

Reply:
Yeah, we probably do want the space, so the user can keep typing their next thing, but that trailing space is not part of the syntax itself, so isn't appropriate for a function with this name.